src/Microsoft.AspNetCore.HostFiltering/HostFilteringMiddleware.cs

```diff
@@ -0,0 +1,175 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Hosting.Server.Features;
+using Microsoft.AspNetCore.Http;
+using Microsoft.AspNetCore.Http.Internal;
+using Microsoft.Extensions.Logging;
+using Microsoft.Extensions.Options;
+using Microsoft.Extensions.Primitives;
+using Microsoft.Net.Http.Headers;
+
+namespace Microsoft.AspNetCore.HostFiltering
+{
+    /// <summary>
+    /// A middleware used to filter requests by their Host header.
+    /// </summary>
+    public class HostFilteringMiddleware
+    {
+        // Matches Http.Sys.
+        private static readonly byte[] DefaultResponse = Encoding.ASCII.GetBytes(
```

Comment:
Ugh, since when do we hard code error pages into our middleware? Shouldn't that be handled separately? What if the client is an API client? Why isn't the 400 enough?

Reply:
There are comments about this elsewhere in the code. This is more severe than your usual failure, it's a binding/protocol level error that has security implications. If you don't provide a body then components like the StatusCode middleware can still generate insecure content. This is one step short of calling Abort.

As for WebAPIs, for an error this severe being able to consume the body is our last concern. That said, I did forget to set the content-type.

Reply:
@shirhatti

Reply:
Gotta agree with Chris here. Best to do what http.sys does